cedar-policy/src/ffi/utils.rs

```diff
@@ -259,84 +259,234 @@ impl From<serde_json::Value> for Entities {
+            Self::Json(json) => crate::Template::from_json(id, json.into())
+                .wrap_err(format!("failed to parse template{msg} from JSON")),
+        }
+    }
+
+    /// Parse a [`Template`] into a [`crate::Template`] and add it into the
+    /// provided [`crate::PolicySet`].
+    pub(super) fn parse_and_add_to_set(
+        self,
+        id: Option<PolicyId>,
+        policies: &mut crate::PolicySet,
+    ) -> Result<(), miette::Report> {
+        let msg = id
+            .clone()
+            .map(|id| format!(" with id `{id}`"))
+            .unwrap_or_default();
+        let template = self.parse(id)?;
+        policies
+            .add_template(template)
+            .wrap_err(format!("failed to add template{msg} to policy set"))
+    }
+}
+
+/// Represents a set of static policies
+#[derive(Debug, Serialize, Deserialize)]
+#[cfg_attr(feature = "wasm", derive(tsify::Tsify))]
+#[cfg_attr(feature = "wasm", tsify(into_wasm_abi, from_wasm_abi))]
+#[serde(untagged)]
+#[serde(
+    expecting = "expected a static policy set represented by a string, JSON array, or JSON object (with no duplicate keys)"
+)]
+pub enum StaticPolicySet {
+    /// Multiple policies as a concatenated string
     Concatenated(String),
-    /// provides multiple policies as a hashmap where the policyId is the key
+    /// Multiple policies as a set
+    Set(Vec<Policy>),
+    /// Multiple policies as a hashmap where the policy id is the key
     #[serde(with = "::serde_with::rust::maps_duplicate_key_is_error")]
-    Map(HashMap<String, String>),
+    Map(HashMap<PolicyId, Policy>),
 }
 
-fn parse_policy_set_from_individual_policies(
-    policies: &HashMap<String, String>,
-    templates: Option<HashMap<String, String>>,
-) -> Result<crate::PolicySet, Vec<miette::Report>> {
-    let mut policy_set = crate::PolicySet::new();
-    let mut errs = Vec::new();
-    for (id, policy_src) in policies {
-        match Policy::parse(Some(id.clone()), policy_src)
-            .wrap_err_with(|| format!("failed to parse policy with id `{id}`"))
-        {
-            Ok(p) => match policy_set
-                .add(p)
-                .wrap_err_with(|| format!("failed to add policy with id `{id}` to policy set"))
-            {
-                Ok(()) => {}
-                Err(e) => {
-                    errs.push(e);
+impl StaticPolicySet {
+    /// Parse a [`StaticPolicySet`] into a [`crate::PolicySet`]
+    pub(super) fn parse(self) -> Result<crate::PolicySet, Vec<miette::Report>> {
+        match self {
+            Self::Concatenated(str) => crate::PolicySet::from_str(&str)
+                .wrap_err("failed to parse policies from string")
+                .map_err(|e| vec![e]),
+            Self::Set(set) => {
+                let mut errs = Vec::new();
+                let policies = set
+                    .into_iter()
+                    .map(|policy| policy.parse(None))
+                    .filter_map(|r| r.map_err(|e| errs.push(e)).ok())
```

Comment:
Would it be more idiomatic here to use something like `.partition` instead of mutating a vec?

Reply:
Possibly 🙂 The examples here https://doc.rust-lang.org/rust-by-example/error/iter_result.html show how to do this in a couple different ways (incl. using `partition`), but I think the `filter_map` + `push` approach is the most readable.